ocaml/networkd/bin/network_server.ml

```diff
@@ -35,19 +35,106 @@ let write_config () =
     try Network_config.write_config !config
     with Network_config.Write_error -> ()
 
+let get_index_from_ethx name =
+  try Scanf.sscanf name "eth%d%!" Option.some with _ -> None
+
+let sort_based_on_ethx () =
+  Sysfs.list ()
+  |> List.filter_map (fun name ->
+         if Sysfs.is_physical name then
+           get_index_from_ethx name |> Option.map (fun i -> (name, i))
+         else
+           None
+     )
+
+let read_previous_inventory () =
+  let previous_inventory = "/var/tmp/.previousInventory" in
+  try
+    Xapi_stdext_unix.Unixext.file_lines_fold
+      (fun acc line ->
+        match Inventory.parse_inventory_entry line with
+        | Some ("MANAGEMENT_INTERFACE", iface) ->
+            info "get management interface from previous inventory: %s" iface ;
+            (iface, snd acc)
+        | Some ("MANAGEMENT_ADDRESS_TYPE", addr_type) ->
+            info "get management address type from previous inventory: %s"
+              addr_type ;
+            (fst acc, addr_type)
+        | _ ->
+            acc
+      )
+      ("", "") previous_inventory
+  with e ->
+    error "Failed to read previous inventory %s: %s" previous_inventory
+      (Printexc.to_string e) ;
+    raise
+      (Network_error
+         (Internal_error
+            (Printf.sprintf "Failed to read previous inventory: %s"
+               previous_inventory
+            )
+         )
+      )
+
+let update_inventory () =
+  let iface, addr_type = read_previous_inventory () in
+  Network_config.write_manage_iface_to_inventory iface addr_type
+
+let handle_upgrade () =
+  let interface_order =
+    match Network_device_order.sort [] with
+    | Ok (interface_order, _) ->
+        interface_order
+    | Error err ->
+        error "Failed to sort interface order [%s]"
+          (Network_device_order.string_of_error err) ;
+        raise (Network_error (Internal_error "sorting failed"))
```

Comment:
This part has been implemented in `do_sort`. The `handle_upgrade` could be `changes_after_upgrade interface_order` to return the changes only.